Index: scripts/ssl-enum-ciphers.nse
===================================================================
--- scripts/ssl-enum-ciphers.nse	(revision 34821)
+++ scripts/ssl-enum-ciphers.nse	(working copy)
@@ -148,9 +148,10 @@
 local function get_record_iter(sock)
   local buffer = ""
   local i = 1
+  local fragment
   return function ()
     local record
-    i, record = tls.record_read(buffer, i)
+    i, record = tls.record_read(buffer, i, fragment)
     if record == nil then
       local status, err
       status, buffer, err = tls.record_buffer(sock, buffer, i)
@@ -157,11 +158,12 @@
       if not status then
         return nil, err
       end
-      i, record = tls.record_read(buffer, i)
+      i, record = tls.record_read(buffer, i, fragment)
       if record == nil then
         return nil, "done"
       end
     end
+    fragment = record.fragment
     return record
   end
 end
